### DISCUSSION TITLE
Annotations for blocks that doesn't contain code.

### DISCUSSION THREAD

**Discussion by UltraInstinct05** [squidfunk/mkdocs-material#3832] [Q&A] [answered]
Is it possible to add code annotations feature (https://squidfunk.github.io/mkdocs-material/reference/code-blocks/#adding-annotations) to markdown fenced blocks that are not *language blocks*

As an example, take the following situation. Here the following is just wrapped in ``` but it doesn't define a language, but I'd still like to use code annotations (it's not code) to describe some information. The problem is, it just doesn't define any language, so there is no concept of comments.

```
.
└── dir_a/
    ├── dir_b # Add a code annotation here describing something special, but # is not a comment char.
    ├── file_1.py
    └── file_2.py
```

**Comment by UltraInstinct05** [comment, marked as answer]
Looks like I just forgot to read the footnotes :) Seems to be on the roadmap https://squidfunk.github.io/mkdocs-material/reference/code-blocks/#fn:1

**Reply by squidfunk** [reply]
Jup, it's on the roadmap but currently not a priority as most issues can be worked around by using a different lexer. For example, you can just use the `sh` lexer in your example, like in the official documentation:

``` sh
.
├─ .icons/                             # Bundled icon sets
├─ assets/
│  ├─ images/                          # Images and icons
│  ├─ javascripts/                     # JavaScript files
│  └─ stylesheets/                     # Style sheets
├─ partials/
│  ├─ integrations/                    # Third-party integrations
│  │  ├─ analytics/                    # Analytics integrations
│  │  └─ analytics.html                # Analytics setup
│  ├─ languages/                       # Translation languages
│  ├─ content.html                     # Page content
│  ├─ ...
│  ├─ toc.html                         # Table of contents
│  └─ toc-item.html                    # Table of contents item
├─ 404.html                            # 404 error page
├─ base.html                           # Base template
└─ main.html                           # Default page
```

This will allow to print directory views with comments, so you can also use code annotations.

<img width="702" alt="Bildschirmfoto 2022-04-22 um 17 32 41" src="https://user-images.githubusercontent.com/932156/164746834-ffe72a84-7fe4-48a9-9e59-a00a550b3180.png">

**Reply by UltraInstinct05** [reply]
Thanks for the alternative example !

**Reply by ewels** [reply]
Quick +1 vote for continuing work on this 🙋🏻 

My additional use case is wanting to add comments within multi-line strings.

For reference, I've just been porting all the training material from asciidocs which does support arbitrary placement, in case that's useful ([docs](https://docs.asciidoctor.org/asciidoc/latest/verbatim/callouts/)).

<details>
<summary>Description of my example use case</summary>

I'm writing docs for a [scientific workflow manager](https://nextflow.io/) and our code blocks typically look like this:

````markdown
```nextflow
process { // (1)
    input: foo
    output: bar
    """
    really-long-multiline-string
    This is a code block in itself, typically a bunch of bash commands
    Lots of stuff going on here, lots to talk about on each line // (2)
    To pygments, it's one big string so impossible to put any comment in the middle
    """
}
```
1. This is regular Nextflow / groovy code, comments work fine here
2. This annotation won't work :(
````

With syntax highlighting to make the lack of `// (2)` comment clearer:

```nextflow
process { // (1)
    input: foo
    output: bar
    """
    really-long-multiline-string
    This is a code block in itself, typically a bunch of bash commands
    Lots of stuff going on here, lots to talk about on each line // (2)
    To pygments, it's one big string so impossible to put any comment in the middle
    """
}
```

We also have quite a lot of console outputs which I really want to leave verbatim from the actual tool output. Plus specifying as bash etc gives a tonne of incorrect syntax highlighting that looks wrong, eg:

```bash
ERROR ~ Error executing process > 'index'

Caused by: 
  Process `index` terminated with an error exit status (127)

Command executed: 

  salmon index --threads 1 -t transcriptome.fa -i index

Command exit status: 
  127

Command output: 
  (empty)

Command error: 
  .command.sh: line 2: salmon: command not found

Work dir: 
  /Users/pditommaso/work/0b/b59f362980defd7376ee0a75b41f62
```

</details>

**Reply by squidfunk** [reply]
AsciiDoc has a fundamentally different implementation. As far as I understand, they are parsing annotations during the build, while we're doing it during runtime. We're relying on Pygments for parsing and mount code annotations in the browser, so the application doesn't get to see the raw source code but what's been highlighted. This means that markers might be destroyed, as parts of them are now wrapped in `span`. Our current implementation requires that code annotations are hosted inside comments, which is the safest thing to do. I don't expect this to happen anytime soon, as we would probably run into too many edge cases, and that's a maintenance nightmare.

**Reply by ewels** [reply]
Gotcha, yeah I absolutely understand 👍🏻 

Code blocks with no syntax, or multi-line strings, should both be pretty sparse in terms of tags. Could there be an `unsafe_annotations` flag for the code fence that allows the code to search somewhat more blindly for the tags? Then people at least know what they're getting into if it doesn't work.

If too nightmarish, I guess I could hack together an override for it 🤔  (`src/assets/javascripts/components/content/annotation/list/index.ts` I guess?)

**Reply by squidfunk** [reply]
Yes, code annotations are detected here:

https://github.com/squidfunk/mkdocs-material/blob/a6565f3e1fa5c4330041a36935dfd850bbd9c38b/src/assets/javascripts/components/content/annotation/list/index.ts#L72-L104

You probably would need to fundamentally alter this functionality, as it currently operates on text nodes. For example, consider the following JavaScript, which showcases the issue of annotation markers being broken into separate lexemes:

``` js
const foo = (1) "string";
```

Result with default configuration:

``` html
<code>
  <span class="kd">const</span>
  <span class="w"> </span>
  <span class="nx">foo</span>
  <span class="w"> </span>
  <span class="o">=</span>
  <span class="w"> </span>
  <!-- Code annotation start -->
  <span class="p">(</span>
  <span class="mf">1</span>
  <span class="p">)</span>
  <!-- Code annotation end -->
  <span class="w"> </span>
  <span class="s2">"string"</span>
  <span class="p">;</span>
</code>
```

Maybe this better illustrates why we've decided to scope annotations to comments. You would need separate preprocessors for each language, and it is hard to decide whether a code annotation is actually an annotation or a language construct. One might argue that one could find a more discernible syntax than `(n)`, but  there's probably no one-size-fits-all solution for all languages.

Another idea might be to somehow pre-process code blocks and extract annotations before passing them to Pygments. I'd expect that to be even more challenging, since you would need to somehow source map back to the original locations, given that Pygments eats strings and returns HTML.

All of that just from the top of my head, and as always happy to be proven wrong.

**Reply by ewels** [reply]
Thanks for the pointer! Yeah I totally understand about the syntax highlighting breaking the strings, I had a similar thought about the difficulty of remapping pre- and post-pygments. Sounds rough.

To be clear, I really like the current comment approach and am not suggesting a change to this. It works beautifully 99% of the time and I agree with the philosophy behind it as well as the implementation logistics.

My train of thought was to have this unsafe flag which would be non-default and would not work for the syntax highlighted case above. It could just loop through all DOM elements, thus be able to catch comments within strings / non highlighted code / multi-line comments and so on (the edge cases where the default doesn't work, but which all should reside within fairly simple HTML).

Arguably not pretty and adds maintenance burden / complexity to have a second method (would maybe just be a case of a flag changing the element selectors? Need to brush up on my typescript). But between the two methods, users could hopefully cover most / all cases.

**Reply by squidfunk** [reply]
I see, so only highlighting plain text, okay. It would be impossible to use `(n)` literals in your code blocks then, even if that's opt-in, since there are no distinctive lexemes in plain text. Regardless of this limitation, yes, that could work, but it's nothing I'd feel comfortable to commit maintaining, as users will definitely start asking for syntax highlighting on those literal blocks.

If you come up with a general method that works for all languages and allows to place markers arbitrarily while retaining syntax highlighting capabilities, I'm definitely interested.

**Reply by ewels** [reply]
Only other method I can think of which is _completely_ general is to stay out of the code block altogether and do everything with line numbers.

For example, maybe:

````md
```txt linenums="1"
Some code
Block here
And stuff
```
* (1) Annotation on first line
* (3) Another annotation on third line
````

**Reply by squidfunk** [reply]
I also already thought of adding that functionality, but that would only allow to _append_ code annotations, not insert them arbitrarily, which is why for now I've not considered implementing this approach.

**Reply by ewels** [reply]
Could always have an optional character start point as well, using a `:` character?

````md
```txt linenums="1"
Some code
Block here
And stuff
```
* (1) Annotation on first line (appended)
* (3:4) Another annotation on third line (starting after "And")
````

**Reply by squidfunk** [reply]
With [Insiders 4.32.0](https://squidfunk.github.io/mkdocs-material/insiders/changelog/#4.32.0), code annotations can now be placed anywhere. Material for MkDocs can be configured per-language to allow for code annotations in other lexemes than comments, e.g. strings. See the documentation on [custom selectors](https://squidfunk.github.io/mkdocs-material/reference/code-blocks/#custom-selectors).

**Reply by ewels** [reply]
Awesome - thank you! 👏🏻🎉

**Comment by UltraInstinct05** [comment]
Is this a bug in code annotations ? I have this example (Windows 11, Firefox Nightly, Google Chrome)

```
    ```sh
    dir_a/
    ├── dir_b/
    │   ├── file_1.py # (1)
    │   ├── file_2.py # (2)
    │   └── file_3.py # (2)
    ├── dir_c/
    └── ...other directories
    ```

    1. Code annotation example 1.
    2. Code annotation example 2.
```

Annotation 1 renders properly

![image](https://user-images.githubusercontent.com/34304644/164754408-74597847-8ff9-409d-90bc-d0487770d140.png)

Annotation 2 doesn't work

![image](https://user-images.githubusercontent.com/34304644/164754471-673d1688-05f9-4ba8-a8d5-6284606ae9a2.png)

Annotation 3 renders properly

![image](https://user-images.githubusercontent.com/34304644/164754517-c761d0a2-9494-470b-a70c-45a43ff446e9.png)

**Reply by UltraInstinct05** [reply]
It looks like I have to define separate numbers for it to not break, which actually breaks re usability.

```
    1. Code annotation example 1.
    2. Code annotation example 2.
    3. Code annotation example 2.
```

**Reply by squidfunk** [reply]
Annotations can only be referenced once, not twice. The reason for this is that the content of the annotation is moved out of the document tree right into the tooltip. No copying or anything is involved. When printed, the content is moved back into the list, so annotations are printable. In order to allow multiple uses of the same annotation, we would need to clone the node, which might lead to unpredictable results.
